### ISSUE TITLE
🐞 - A tree renders all children(and children all nested children) event if parents are not expanded.

### ISSUE THREAD

**Issue by EricPoul** [Tinkoff/taiga-ui#1665]
### Which @taiga-ui/* package(s) are the source of the bug?

kit

### Description

The tree should render children only if it's expanded. If my trees are made of an array of 3000 items, it will leak a lot of memory and a colossal performance downgrade. Especially if I render it in a dropdown.

I guess the fix is as simple as it gets.

We need to add `ngIf` in `tree.component.html`

```html
<tui-tree-item
    *tuiLet="children$ | async as children"
    #view
    [tuiTreeNode]="value"
    [expandable]="isExpandable"
>
    <div
        *ngIf="value !== children"
        polymorpheus-outlet
        [content]="content"
        [context]="{$implicit: value, node: view}"
    ></div>

     <ng-container *ngIf="item?.isExpanded" ngProjectAs="tui-tree">
        <tui-tree
           *ngFor="let child of children"
           [value]="child"
            [content]="content"
        ></tui-tree>
    </ng-container>
</tui-tree-item>
```

add a getter in `tree.component.ts` (because we have a handler to check whether it is expandable or not)

```ts
get isExpandable(): boolean {
    return !!this.handler(this.value).length;
  }
```

add an input to the `tree-item.component.ts` and remove `@ContentChildren nested` due to its needless.

Also I'd like to have an ability to pass `trackBy` to nested trees.

### Angular version

13.1.3

### Taiga UI version

2.34.0

### Which browsers have you used?

- [X] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux
- [ ] iOS
- [ ] Android

**Comment by waterplea** [comment]
1. It would probably still work since after `ngIf` turn to `true` it would have `tui-tree` inside content. Give it a shot 🙂
2. Yeah, probably. With `@Optional`.
3. Browser will not re-render children if they are equal by `===` even without `trackBy`. If you update your array, say, by `items = items.filter(...)` then it would reuse the same DOM elements and components it had for the items that remain in the array. You would only need `trackBy` if you reload the same items from the server request, for example, and the same objects in the array will not be equal to each other — then you could artificially tell Angular that they are considered equal if, for example, they have the same `id` property.

**Comment by EricPoul** [comment]
Yeah, you described it more correctly than I did. Anyway, it was just a suggestion for optimization. In most of the cases, it'll be excess, not everyone has 3k items in a tree.😄

**Comment by EricPoul** [comment]
It doesn't work without `ngProjectAs`. But we can make the next 
```ts
 readonly children$ = this.check$.pipe(
        startWith(null),
        // was map(() => this.handler(this.value)),
        map(() => this.isExpandable ? this.handler(this.value) : []),
        distinctUntilChanged(),
    );
```

**Comment by waterplea** [comment]
If it works I like it better :)

**Comment by waterplea** [comment]
I believe you meant `this.isExpanded`?

**Comment by EricPoul** [comment]
it's a new `getter` that shows whether it has children or not

**Comment by waterplea** [comment]
But it only returns true when there are children and false when there are not, the ternary doesn't make sense in this case 🤔

**Comment by EricPoul** [comment]
We need this getter in `tui-tree-item` as well 
```ts
 @HostBinding('class._expandable')
    get isExpandable(): boolean {
        return !!this.parent?.isExpandable;
    }
```

**Comment by waterplea** [comment]
Go ahead with the PR and we will continue discussion there 🙂

**Comment by shiv9604** [comment]
Hi @waterplea, Can you please confirm If the issue is resolved? I can see PR's for it but not sure if we have any more unresolved tasks for this so I can work on it.

**Comment by waterplea** [comment]
> Hi @waterplea, Can you please confirm If the issue is resolved? I can see PR's for it but not sure if we have any more unresolved tasks for this so I can work on it.

This issue is not resolved, the linked ones are different, just a bit related

**Comment by shiv9604** [comment]
Hi @waterplea,

I’ve debugged the issue and identified the following behavior:
- When the `tuiTreeController` is passed `true` in `<tui-tree/>`, all child nodes are rendered regardless of the parent node's state.
- If it’s set to `false`, the children are only rendered when the parent is expanded, which is user-configurable.

**Suggested Improvement:**
To enhance clarity, I suggest renaming `tuiTreeController` to something more descriptive, such as `expanded`. This would clearly indicate that the property controls whether the tree nodes are expanded by default.

**Regarding Documentation:**
If the goal is to have the tree in a collapsed state by default, we can update the documentation accordingly. Please let me know if the desired behavior in the tree component differs, or if there are specific aspects in the `kit` package that need further investigation or handling within the component.
